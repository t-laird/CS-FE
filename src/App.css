.App {
  min-height: 100vh;
  background: #273847;
  display: grid;
  grid-template-rows: 80px 150px auto; }
  .App header {
    background-color: #3B5369;
    height: 60px;
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 10px 80px;
    margin-bottom: 80px; }
    .App header img {
      height: 40px; }
    .App header h1 {
      margin: 0px;
      color: white; }
  .App .showUpload {
    background: white;
    border: 1px solid black;
    padding: 10px;
    margin: 50px auto;
    display: block; }
