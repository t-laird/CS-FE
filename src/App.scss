.App {
  min-height: 100vh;
  background: #273847;
  display: grid;
  grid-template-rows: 80px 150px auto;
  header {
    background-color: #3B5369;
    height: 60px;
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 10px 80px;
    margin-bottom: 80px;
    img {
      height: 40px;
    }
    h1 {
      margin: 0px;
      color: white;
    }
  }

  .showUpload {
    background: white;
    border: 1px solid black;
    padding: 10px;
    margin: 50px auto;
    display: block;
  }
}