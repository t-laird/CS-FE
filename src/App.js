import React, { Component } from 'react';
import './App.css';
import UploadForm from './Components/UploadForm/UploadForm';
import SiteAvailability from './Components/SiteAvailability/SiteAvailability';
import logo from'./cs-logo.svg';
import moment from 'moment';


class App extends Component {
  constructor() {
    super();
    this.state = {
      campsites: {},
      reservations: [],
      search: {}
    };
  }

  async componentDidMount() {
    const testEndpoint = await fetch('/campsites');
    const data = await testEndpoint.json();
    console.log(testEndpoint);
    console.log(data);

  }

  checkAvailability = (start, end, resSize, campsites, res) => {
    for (let i = 0 ; i < 6 ; i++) {
      if (
        (
          moment(start).diff(end, 'days') !== 1 &&
          moment(start).diff(end, 'days') <= i + 1 &&
          moment(start).diff(end, 'days') > 0
        ) 
        || 
        (
          moment(start).diff(end, 'days') <= 0 && 
          moment(start).diff(end, 'days') >= (resSize * -1)
        )
      ) {
        campsites[res.campsiteId].gaps[i] = false;
      }
    }
    return campsites;
  }

  generateAvailability = (reservations, campsites, search) => {
    if (this.state.dataDigested) return;

    let campsitesCopy = Object.assign({}, campsites);

    const startDate = moment(search.startDate);
    const endDate = moment(search.endDate);

    const reservationSize = endDate.diff(startDate, 'days') + 1;

    reservations.forEach( (res, ind) => {
      campsitesCopy = this.checkAvailability(res.startDate, endDate, reservationSize, campsitesCopy, res, this.state.gap);

      campsitesCopy = this.checkAvailability(startDate, res.endDate, reservationSize, campsitesCopy, res, this.state.gap);
    });

    this.setState({
      search, campsites: campsitesCopy, reservations
    });
  }

  updateData = (data) => {
    const { search, campsites, reservations } = data;

    const campsitesToObj = campsites.reduce((res, campsite) => {
      res[campsite.id] = {};
      res[campsite.id].name = campsite.name;
      res[campsite.id].gaps = {
        0: true,
        1: true,
        2: true,
        3: true,
        4: true,
        5: true,
      };
      return res;
    }, {});

    return this.generateAvailability(reservations, campsitesToObj, search);
  }

  render() {
    return (
      <div className="App">
        <header className="App-header">
          <img src={logo} alt="Campspot Logo" />
          <h1>Code Challenge</h1>
        </header>
       <UploadForm updateData={this.updateData}/>
       <SiteAvailability 
        reservations={this.state.reservations} 
        campsites={this.state.campsites} 
        search={this.state.search} />
      </div>
    );
  }
}

export default App;
