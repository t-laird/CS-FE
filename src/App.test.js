import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow } from 'enzyme';

describe('App tests', () => {
  let renderedApp;
  beforeEach(() => {
    renderedApp = shallow(<App />);
  });
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should match the snapshot', () => {
    expect(renderedApp).toMatchSnapshot();
  });

  it('should render the correct default components', () => {
    expect(renderedApp.find('header').length).toEqual(1);
    expect(renderedApp.find('UploadForm').length).toEqual(1);
    expect(renderedApp.find('SiteAvailability').length).toEqual(1);
  });

  it('should set state with the formatted received data when a valid JSON file is uploaded', () => {
    const mockData = { 
      search: {startDate: "2016-06-04", endDate: "2016-06-06"}, 
      reservations: [{ campsiteId: 1, startDate: "2016-06-01", endDate: "2016-06-02" }],
      campsites: [{
        "id": 1,
        "name": "Grizzly Adams Adventure Cabin"
      }]
    };

    const formattedData =   {
      "campsites": { "1": { "name": "Grizzly Adams Adventure Cabin", "gaps": { "0": true, "1": false, "2": false, "3": false, "4": false, "5": false } } },
      "reservations": [ { "campsiteId": 1, "startDate": "2016-06-01", "endDate": "2016-06-02" }],
      "search": { "startDate": "2016-06-04", "endDate": "2016-06-06" },
    };

    renderedApp.instance().updateData(mockData);

    expect(renderedApp.state('campsites')).toEqual(formattedData.campsites);
    expect(renderedApp.state('reservations')).toEqual(formattedData.reservations);
    expect(renderedApp.state('search')).toEqual(formattedData.search);
  });

  it('Should return all false for a search overlapping an existing reservation', () => {
    const mockData = { 
      search: {startDate: "2016-06-04", endDate: "2016-06-06"}, 
      reservations: [{ campsiteId: 1, startDate: "2016-06-01", endDate: "2016-06-04" }],
      campsites: [{
        "id": 1,
        "name": "Grizzly Adams Adventure Cabin"
      }]
    };
    
    const expectedCampsiteData = {
      "1": { "name": "Grizzly Adams Adventure Cabin", "gaps": { "0": false, "1": false, "2": false, "3": false, "4": false, "5": false } }
    };
    
    renderedApp.instance().updateData(mockData);
    expect(renderedApp.state().campsites).toEqual(expectedCampsiteData);
  });

  it('Should return false for all but 0 gap given a reservation with a 1 day gap after the search', () => {
    const mockData = { 
      search: {startDate: "2016-06-04", endDate: "2016-06-06"}, 
      reservations: [{ campsiteId: 1, startDate: "2016-06-08", endDate: "2016-06-10" }],
      campsites: [{
        "id": 1,
        "name": "Grizzly Adams Adventure Cabin"
      }]
    };
    
    const expectedCampsiteData = {
      "1": { "name": "Grizzly Adams Adventure Cabin", "gaps": { "0": true, "1": false, "2": false, "3": false, "4": false, "5": false } }
    };
    
    renderedApp.instance().updateData(mockData);
    expect(renderedApp.state().campsites).toEqual(expectedCampsiteData);
  });

  it('Should return false for all but 0 gap given a reservation with a 1 day gap before the search', () => {
    const mockData = { 
      search: {startDate: "2016-06-04", endDate: "2016-06-06"}, 
      reservations: [{ campsiteId: 1, startDate: "2016-05-30", endDate: "2016-06-02" }],
      campsites: [{
        "id": 1,
        "name": "Grizzly Adams Adventure Cabin"
      }]
    };
    
    const expectedCampsiteData = {
      "1": { "name": "Grizzly Adams Adventure Cabin", "gaps": { "0": true, "1": false, "2": false, "3": false, "4": false, "5": false } }
    };
    
    renderedApp.instance().updateData(mockData);
    expect(renderedApp.state().campsites).toEqual(expectedCampsiteData);
  });

  it('Should return false for all but 0 and 1 gap given a reservation with a 2 day gap before the search', () => {
    const mockData = { 
      search: {startDate: "2016-06-04", endDate: "2016-06-06"}, 
      reservations: [{ campsiteId: 1, startDate: "2016-05-30", endDate: "2016-06-01" }],
      campsites: [{
        "id": 1,
        "name": "Grizzly Adams Adventure Cabin"
      }]
    };
    
    const expectedCampsiteData = {
      "1": { "name": "Grizzly Adams Adventure Cabin", "gaps": { "0": true, "1": true, "2": false, "3": false, "4": false, "5": false } }
    };
    
    renderedApp.instance().updateData(mockData);
    expect(renderedApp.state().campsites).toEqual(expectedCampsiteData);
  });

  it('Should return false for all but 0 and 1 gap given a reservation with a 2 day gap after the search', () => {
    const mockData = { 
      search: {startDate: "2016-06-04", endDate: "2016-06-06"}, 
      reservations: [{ campsiteId: 1, startDate: "2016-06-09", endDate: "2016-06-12" }],
      campsites: [{
        "id": 1,
        "name": "Grizzly Adams Adventure Cabin"
      }]
    };
    
    const expectedCampsiteData = {
      "1": { "name": "Grizzly Adams Adventure Cabin", "gaps": { "0": true, "1": true, "2": false, "3": false, "4": false, "5": false } }
    };
    
    renderedApp.instance().updateData(mockData);
    expect(renderedApp.state().campsites).toEqual(expectedCampsiteData);
  });

  it('Should return false for all but 0, 1 and 2 gap given a reservation with a 3 day gap after the search', () => {
    const mockData = { 
      search: {startDate: "2016-06-04", endDate: "2016-06-06"}, 
      reservations: [{ campsiteId: 1, startDate: "2016-06-10", endDate: "2016-06-13" }],
      campsites: [{
        "id": 1,
        "name": "Grizzly Adams Adventure Cabin"
      }]
    };
    
    const expectedCampsiteData = {
      "1": { "name": "Grizzly Adams Adventure Cabin", "gaps": { "0": true, "1": true, "2": true, "3": false, "4": false, "5": false } }
    };
    
    renderedApp.instance().updateData(mockData);
    expect(renderedApp.state().campsites).toEqual(expectedCampsiteData);
  });

  it('Should return false for all but 0 and 1 gap given a reservation with a 2 day gap before the search', () => {
    const mockData = { 
      search: {startDate: "2016-06-04", endDate: "2016-06-06"}, 
      reservations: [{ campsiteId: 1, startDate: "2016-05-27", endDate: "2016-05-31" }],
      campsites: [{
        "id": 1,
        "name": "Grizzly Adams Adventure Cabin"
      }]
    };
    
    const expectedCampsiteData = {
      "1": { "name": "Grizzly Adams Adventure Cabin", "gaps": { "0": true, "1": true, "2": true, "3": false, "4": false, "5": false } }
    };
    
    renderedApp.instance().updateData(mockData);
    expect(renderedApp.state().campsites).toEqual(expectedCampsiteData);
  });

  it('Should return true for all given a reservation adjacent to the end date of the search', () => {
    const mockData = { 
      search: {startDate: "2016-06-04", endDate: "2016-06-06"}, 
      reservations: [{ campsiteId: 1, startDate: "2016-06-07", endDate: "2016-06-10" }],
      campsites: [{
        "id": 1,
        "name": "Grizzly Adams Adventure Cabin"
      }]
    };
    
    const expectedCampsiteData = {
      "1": { "name": "Grizzly Adams Adventure Cabin", "gaps": { "0": true, "1": true, "2": true, "3": true, "4": true, "5": true } }
    };
    
    renderedApp.instance().updateData(mockData);
    expect(renderedApp.state().campsites).toEqual(expectedCampsiteData);
  });

  it('Should return true for all given a reservation adjacent to the start date of the search', () => {
    const mockData = { 
      search: {startDate: "2016-06-04", endDate: "2016-06-06"}, 
      reservations: [{ campsiteId: 1, startDate: "2016-06-01", endDate: "2016-06-03" }],
      campsites: [{
        "id": 1,
        "name": "Grizzly Adams Adventure Cabin"
      }]
    };
    
    const expectedCampsiteData = {
      "1": { "name": "Grizzly Adams Adventure Cabin", "gaps": { "0": true, "1": true, "2": true, "3": true, "4": true, "5": true } }
    };
    
    renderedApp.instance().updateData(mockData);
    expect(renderedApp.state().campsites).toEqual(expectedCampsiteData);
  });  
});
