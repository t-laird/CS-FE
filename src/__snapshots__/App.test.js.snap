// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="App-header"
>
          <img
                    alt="Campspot Logo"
                    src="cs-logo.svg"
          />
          <h1>
                    Code Challenge
          </h1>
</header>,
        <UploadForm
          updateData={[Function]}
/>,
        <SiteAvailability
          campsites={Object {}}
          reservations={Array []}
          search={Object {}}
/>,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <img
              alt="Campspot Logo"
              src="cs-logo.svg"
/>,
            <h1>
              Code Challenge
</h1>,
          ],
          "className": "App-header",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "Campspot Logo",
              "src": "cs-logo.svg",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Code Challenge",
            },
            "ref": null,
            "rendered": "Code Challenge",
            "type": "h1",
          },
        ],
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "updateData": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "campsites": Object {},
          "reservations": Array [],
          "search": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="App-header"
>
            <img
                        alt="Campspot Logo"
                        src="cs-logo.svg"
            />
            <h1>
                        Code Challenge
            </h1>
</header>,
          <UploadForm
            updateData={[Function]}
/>,
          <SiteAvailability
            campsites={Object {}}
            reservations={Array []}
            search={Object {}}
/>,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <img
                alt="Campspot Logo"
                src="cs-logo.svg"
/>,
              <h1>
                Code Challenge
</h1>,
            ],
            "className": "App-header",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "Campspot Logo",
                "src": "cs-logo.svg",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Code Challenge",
              },
              "ref": null,
              "rendered": "Code Challenge",
              "type": "h1",
            },
          ],
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "updateData": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "campsites": Object {},
            "reservations": Array [],
            "search": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
